From ef9b72306128325c282358bacff32f03c22fe827 Mon Sep 17 00:00:00 2001
From: Nicholas Sim <nicholassimws@gmail.com>
Date: Mon, 12 Feb 2018 14:34:11 +0000
Subject: [PATCH 09/25] dle: blocking: fix bounds on intra loop

---
 devito/dle/backends/advanced.py | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/devito/dle/backends/advanced.py b/devito/dle/backends/advanced.py
index e4d628aa..710170ac 100644
--- a/devito/dle/backends/advanced.py
+++ b/devito/dle/backends/advanced.py
@@ -8,7 +8,7 @@ from itertools import combinations
 import cgen
 import numpy as np
 import psutil
-from sympy import Min
+from sympy import Min, Max
 
 from devito.cgen_utils import ccode
 from devito.dimension import Dimension
@@ -164,11 +164,11 @@ class DevitoRewriter(BasicRewriter):
 
                 inter_block = Iteration([], dim, [start, finish, block_size],
                                         properties=PARALLEL)
-                inter_blocks.append(inter_block)
+                inter_blocks.append(inter_block)  # the area being blocked
 
                 # Build Iteration within a block
-                start = inter_block.dim
-                finish = Min(start + block_size, finish)  # FIXME: "widen", FIXME: + eps?
+                start = Max(inter_block.dim, start)
+                finish = Min(inter_block.dim + block_size, finish)
                 intra_block = i._rebuild([], limits=[start, finish, 1], offsets=None,
                                          properties=i.properties + (TAG, ELEMENTAL))
                 intra_blocks.append(intra_block)
-- 
2.17.1

