From 246b8e10113b8cb8fefe8b6a0952a867bc16a9bd Mon Sep 17 00:00:00 2001
From: Nicholas Sim <nicholassimws@gmail.com>
Date: Mon, 5 Mar 2018 11:06:48 +0000
Subject: [PATCH 19/25] dse: skewing: fix time dimension detection bug

---
 devito/dse/backends/advanced.py | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/devito/dse/backends/advanced.py b/devito/dse/backends/advanced.py
index d060461e..f03e70bd 100644
--- a/devito/dse/backends/advanced.py
+++ b/devito/dse/backends/advanced.py
@@ -2,6 +2,7 @@ from __future__ import absolute_import
 
 from collections import OrderedDict
 
+from devito.dimension import TimeDimension
 from devito.ir import clusterize
 from devito.dse.aliases import collect
 from devito.dse.backends import BasicRewriter, dse_pass
@@ -186,7 +187,10 @@ class SkewingRewriter(AdvancedRewriter):
                 mapper[dim] = dim + skew_factor * t
                 skews[dim] = (skew_factor, t)
             elif dim.is_Time:
-                t = dim.parent
+                if isinstance(dim, TimeDimension):
+                    t = dim
+                elif isinstance(dim.parent, TimeDimension):
+                    t = dim.parent
 
         if t is None:
             return cluster
-- 
2.17.1

